### ISSUE TITLE
Support for namespaces

### ISSUE THREAD

**Issue by tebanep** [michaelsauter/crane#134]
It would be nice if Crane supported namespaces lke fig does. Having an option in the crane.yml file or using its parent directory as prefix could be some options.

Thanks for this great tool, I think it's much better in many ways!

**Comment by bjaglin** [comment]
I've personally worked around the lack of built-in namespaces by prefixing container names declarations and references with an environment variable, and apart from the slightly more verbose `crane.yml`, I am satisfied.

**Comment by tebanep** [comment]
@bjaglin it is a very good workaround, however it would be nice to have this option so that your crane.yml doesn't get polluted with "${VAR}_container_name" everywhere. Thanks for your hint!

**Comment by michaelsauter** [comment]
@tebanep Thanks, glad you like Crane!

The reason why namespaces are not in Crane as of yet is that I tried to design everything to it maps 1:1 to docker commands. Which means, if you container name in `crane.yaml` is `some-container` then you should be able to do `docker stop some-container`. Nothing in Crane should stop you from using plain Docker commands.

That being said, it seems not unreasonable to want namespaces. I think the workaround mentioned by @bjaglin is good enough for this not to be a big issue. However, I would probably not be against some opt-in namespacing, but I'm too busy right now to add it. If you want it, please contribute it :)

Implementation wise, I think what you suggested makes sense. So some option in the root of the config, e.g. `namespace: some_namespace`. Or just `namespace: true` and it uses the folder name. Not sure about these ... just some initial thoughts.

**Comment by tebanep** [comment]
@michaelsauter thanks for your fast response! Though I'm pretty new in the programming field, for sure I'll try my best and try to contribute. Best of luck with all your projects and thanks again for this useful tool. It even has a much better and "dockerized" name than the official one!  ;)

**Comment by michaelsauter** [comment]
Recently, I needed namespace quite a bit and while the environment variable workaround works, it's a bit cumbersome.

I would like to tackle this soonish, maybe next week. I'd like to get some feedback for my initial thoughts:

* Add an optional top-level `namespace` key. It can be either `true` in which case it will append `<foldername>_` to the containers. If it is a string, that will be used instead of the folder name.
* Additionally, I would like to make this configurable via the CLI. My use case is that I have a set of containers described by a crane config. I use that set in CI. In order to run builds in parallel, I would like to pass the commit hash that I'm building as the namespace. Currently, I'm doing this like `COMMIT_SHA=abcdefg crane run --recreate`. It would be nice to use `crane run --recreate --namespace=abcdefg`

Thoughts on this? I realize I can get a cleaner crane config without the second step, by just using the environment variable for the namespace key, but I think it might be nice to have this as an argument.

**Comment by dreamcat4** [comment]
I suppose it can be made flexible with a format pattern like `sprintf()`. I mean in the implementation (not necessary to expose in the cmdline. That was if someone else needs to they can just change the fmt string in your sourcecode to choose whether it's adding a prefix or suffix (before or after).

**Comment by bjaglin** [comment]
:+1: on your initial thoughts. Although I guess that if you add the `--namespace` flag, I don't see a use case for supporting a hardcoded namespace prefix in `crane.yml` (I find the bool or string value ambiguous)?

I would only add a top-level bool `namespace.working-directory` to support the directory use-case (we are talking about the current directory right, not the parent directory of `crane.yml` ?), and leave some room for other namespacing strategies (by day, by user, etc).

**Comment by michaelsauter** [comment]
Thanks for all the feedback!

I think that the feature should be really simple. I don't think there is a need for Crane to support every possible flavour of namespacing one might want. Basically, I just want to enable users to create isolated sets of containers with one crane config.

E.g. I think that prefixing is enough, no need for a suffix. If you wanted to have a suffix, well, you can probably do it with a prefix as well. If not, there always is the environemnt variable fallback.

In regards to working directory vs. parent directory of crane, I was actually thinking parent directory of crane. Working directory is more flexible, but could be quite confusing if you execute Crane from a subdir and are not aware of the implications.

Also, I just realised `-n` is already taking for `--no-cache`. Maybe we can use `--prefix`? Or change the `no-cache` thing, since I don't consider it as valuable.

**Comment by bjaglin** [comment]
I don't see how using the parent directory would serve the goal of being able to start several stacks from the same configuration, since the prefix would always be the same? 

Still not sold on the "bool or string" flag, which for me is confusing. If there is a CLI flag, the "prefixing strategy" (such as directory-based) can be passed via `-n=$(pwd)`, so I find it overkill to code that directly in Crane.

**Comment by michaelsauter** [comment]
Good point. Since I only have a need for the CLI flag right now, I'm going to implement just that. Some configuration option can come later if people want it badly :)

**Comment by michaelsauter** [comment]
Postponed till after 2.0 (see #212).

**Comment by michaelsauter** [comment]
There will be a prefix flag in 2.0, which I think covers this use case here as well (not perfectly, but well enough I think).
